### DISCUSSION TITLE
Multiple identities per account

### DISCUSSION THREAD

**Discussion by gdude2002** [stoatchat/discussions#76] [Feature Suggestions]
This is a pretty specific request, so bear with me.

# Summary

On Discord, there's an ecosystem of bots that exist purely to resend messages using a different username or avatar. Because Discord has no native support for this, they use webhooks - so they appear as follows:

![image](https://user-images.githubusercontent.com/204153/131220492-11a542ca-8263-43f7-9ee8-dc62d195f0d8.png)

This post proposes adding a system for multiple identities per account to Revolt, while the systems are still new and malleable.

## Rationale

There are two major groups I have in mind with this feature.

### 1. Plural systems

For those that aren't aware, there exists a subset of the population that are known as **plural systems**. This term is used to describe people with pluralities - multiple people within one body, so to speak. If you're not familiar with the concept, feel free to take a look at [this primer](https://morethanone.info/). Members of plural systems are sometimes called "alters", if you're already familiar with that term.

Plurality tools like the bots I've mentioned above are a key point of comfort and communication for members of plural systems. They allow each individual member to interact with an avatar and name of their own, which means they're not constrained to a single identity. However, Discord doesn't go far enough, here - there are strict rate limits on nickname and avatar changes (which would already be incredibly burdensome, especially given how fast switches can happen and the fact that there may be multiple members chatting at once), and no built-in identity system.

A properly integrated solution is something Revolt could break ground on here, and it would be a massive bonus for all plural systems.

### 2. Roleplayers

Another common use for bots like those I've mentioned above is for role-playing servers. As you can probably tell, having multiple identities available at the push of a button would be extremely convenient.

### Niche use-cases

I could also envision other use-cases here, such as staff members using separate identities for personal and official communications - I'm sure there are other possibilities.

## Concept

There are many ways this could be exposed in the frontend. I'll leave the configuration side of this to the community and developers, but I've had a couple of ideas on how the UX might look:

![multi-identities](https://user-images.githubusercontent.com/204153/131222035-b80a3d3b-1dd7-4daa-a724-4f4a3b8b6eae.png)

You can probably tell I'm not a graphic designer, haha

**Comment by Yu-Vitaqua-fer-Chronos** [comment]
Personally we like the third option, with multiple identities! There are plenty of times where me and someone else are co-fronting and want to say a similar thing, so we send one message, but it's frustrating where it seems like only one of us is actually talking when we're both wanting to get our say

**Reply by Saragon4005** [reply]
Extending this to also have shared identities sort of like organization accounts would be a niche but appreciated feature, especially for official communications.

**Reply by Tinkerpaw** [reply]
Definitely agree there, as that feature was extremely useful on telegram so administrators could use a shared alias for announcements and the like.

**Reply by Robo-Fortune** [reply]
Very much agreed! This would be immensely helpful.

**Comment by luka-devnull** [comment]
another thing this would probably be immensely useful for: bridge bots. it'd be a lot nicer from the revolt side of things vs abusing webhooks for it because unlike webhooks a bot could create multiple identities that don't have to be reused for every message like a webhook is. also: identities could include better information about where the bridged user is coming from and which bot is doing the bridging without showing it in the main message (also unlike webhooks)

also there's also a whole slew of discord bots that have all kinds of silly minigames and whatnot, can def see huge potential in them being able to spawn new identities for whatever sort of things people can come up with

**Reply by gdude2002** [reply]
That's fascinating, I hadn't considered bots in my original proposal. Strongly agree though, no doubt this would be extremely useful.

**Reply by Saragon4005** [reply]
I've always imagined bridges to be more native then bots but still use a system like webhooks, this would work just as well it not better because this would allow more modularity. It also solves the issue of bots like this being basically unmoderatable. Making a setting for requesting different identities to be treated like different user accounts would help with moderation bots, although it might still be a bit problematic, unless each identity is granted a different ID, maybe based off the original if that doesn't conflict with the spec?

**Comment by onpon4** [comment]
We would very much like this feature too! Would be even better than PluralKit since it could be made to apply to DMs.

🕵️

**Reply by FreshAlacrity** [reply]
Absolutely yes - that would be huge! It would also help with messages bouncing up and down since they wouldn't need to be deleted and reposted, and my system has at least one friend for whom that in particular is a major accessibility issue.

**Reply by WompWomp312** [reply]
> Absolutely yes - that would be huge! It would also help with messages bouncing up and down since they wouldn't need to be deleted and reposted, and my system has at least one friend for whom that in particular is a major accessibility issue.

yeah, that's smth that can get disorienting and make it hard to follow a convo with stuff moving constantly :P

**Comment by selenated** [comment]
+1, have been bouncing around an idea for a discord-like app with this feature for a while now. Having an integrated way for users to post under aliases that still link back to the main account in a clear and easily accessible way is something I've been interested in since fiddling with roleplay forums as a kid, and as a plural person it's *super* handy for us

I'd be willing to help make this happen once I've got time to take a look at everything 👀

**Reply by FreshAlacrity** [reply]
Heck yeah! Hope you get the time to look into it :tada:

**Comment by envoidia** [comment]
This would be huge, it's everything we've ever wanted in a chat app, and would absolutely cement Revolt as the only choice for us and many others. Please, please do this!

**Comment by selenated** [comment]
We commented on this before, but for others interested: this is currently being worked on with the masquerade feature :D

As far as I know it's only available through the API but it should get added into the client as well

**Comment by Saragon4005** [comment]
As someone who has been dealing with systems more and more adding this would solve so many issues. PluralKit breaks so many features bots, search, not role permissions at least,  but it does break visual roles. Having this be integrated into the main chat app would automatically solve all these issues.

**Comment by erkinalp** [comment]
This feature should be tightly coupled with permissions to control which user can use which identities, in a many to many fashion.

**Reply by gdude2002** [reply]
Strongly disagree - if one of the goals of this is to provide an accessibility tool, gatekeeping (at a permissions level) which identities may be used is contradictory.

As long as the UI makes it obvious which account the message is coming from, I think it should be fairly simple to moderate troublemakers, or - when the permissions system is more fleshed out - remove masquerade permissions from them.

**Reply by Rexogamer** [reply]
The ability to use masquerades (which are ultimately the implementation of this) can be controlled per-role/channel by enabling/disabling the Use Masquerades permission.

Masquerades are also user specific - they're not "shared", although nothing is stopping two users with the Use Masquerades permission from using the same name and profile picture.

**Reply by erkinalp** [reply]
@Rexogamer I meant something like user aliases which would allow things like posting a message in behalf of an institution (hence many to manyness), rather than masquerades.

**Reply by Rexogamer** [reply]
That seems, to me at least, to be a separate idea. As far as I can tell, this discussion is about personal masquerades - we may add shared masquerades, but for now shared accounts are the best available option for posting as an organisation and such.

**Comment by Tinkerpaw** [comment]
Reboosting this particular request, as the new username system makes this much easier to do.

**Comment by arisu0** [comment]
boosting this too, hope it gets added soon

**Comment by ghost** [comment]
this would be great! boosting as well

**Comment by Robo-Fortune** [comment]
This would be immensely helpful, I hope it makes it into the new client! If not, I plan to try to implement it myself anyways.

**Comment by sinnom** [comment]
This would be so so helpful to have! Hope this gets implemented.

**Comment by aprildoesthings** [comment]
We would really like to have a feature like this! It would help us when communicating with others so much!

**Comment by lunathir** [comment]
We think this would work very well, would prefer this over Discord's webhook function stuff

**Comment by sinnom** [comment]
Still very excited for this feature! I would also ask that it please be named something else, like "Proxies" or "Profiles" or "Display Profiles" or "Sub-Profiles". There's an unfortunate stereotype of systems and pluralfolk just "faking" or "pretending", and the name "Masquerades" unfortunately affirms that stereotype.

**Comment by Beethoven-n** [comment]
what repo should this be made an issue on?
this organization has 37 repos and i'm unfamiliar

**Reply by erkinalp** [reply]
revoltchat/backend

**Reply by Beethoven-n** [reply]
thanks bunches!

**Comment by viiuan** [comment]
pleeeease please can this be real someday so i can fully abandon discord <3

**Comment by alisticks** [comment]
as a system, HUGE want for this. would be awesome especially since im not aware of any revolt bots for this kind of thing

**Reply by FreshAlacrity** [reply]
https://github.com/Ninty9/RevoltKit has been working well for my system as an immediate/short-term work-around, but of course built-in support would be a huge quality of life improvement against anything else out there :+1:

**Comment by remeranAuthor** [comment]
There are a lot of other pretty good use cases for such a feature in a professional capacity, such as a specific identity for server management announcements that is usable by all of the administrators, or a "serious mode" mod up mode like how IRC has.

**Comment by SixtyEmeralds** [comment]
If someone implemented this, I think my headmate and I could actually, finally commit to using a single account somewhere instead of our kludgy and muddy triggered response of having separate accounts.  I don't want to mislead people as to who we are, it's just we take identity seriously, and don't feel comfortable posting on each other's accounts.

**Comment by and-rea** [comment]
I think when it's two people, it is correct use two separate accounts. You never know when you both go separate ways for any reason and the result is always a mess.
Multiple persona on same account maybe can have its use as some business account, as "support" account as example. But it is better handled with roles IMHO.

If i want my same account appear as two different persona, it is because:
1) privacy: I don't want other people to know I am the same persona in some different channels I prefer to keep alone;
2) roleplay: I can better represent my character in some game going on;

On Discord there is the distinct profiles per server but is behind a paywall and still doesn't solve the issue, as my account is still the same in both serveres even if profiles are dedicated.

**Reply by gdude2002** [reply]
This proposal doesn't really have anything to do with two separate people sharing an account, but realistically I'm sure a small handful of people will find it useful to do so - for example, perhaps parents might prefer to have a single account both their kids use so it's easier to monitor them.

The proposal was meant mostly with [plural systems](https://moderation.wiki/a/plurality) in mind, but I have to say, there are a lot of other potential use-cases too!

**Comment by IAmTomahawkx** [comment]
fwiw this is on our radar, but its unlikely to be coming in the near future due to some other stuff we're already in the works on. Keep up the suggestions/ideas for this.

**Comment by and-rea** [comment]
I had recently a little chat about this on the Revolt lounge, that ended up in a couple considerations that are imho worth mentioning here.

The first is about the question "why have multiple identities under the same account, when multiple accounts are allowed so you can make an account per identity?"
Looking at my own specifics needs (keeping private sets of joined channels, and roleplaying) it is in effect a good solution: if I want things be separated, I want two separate accounts.
On this view, I still have two concerns:
1) moving between accounts is not convenient: I have to go 'thru logoff and login processes. Discord solved this with the "account switch" feature, so I think this can be done here too;
2) premium features: Revolt sooner or later will have to launch monetization to pay the bills. If the premium is tied to the account, I will have to pay multiple times the same premium for the accounts I am using. One reason that wants me go away Discord is having moved their "nitro" features that made me sub there to an higer, more expensive "tier" and I now have canceled the sub. We all know, we pay these subs for fun and convenience, but we can live and breath even without. I surely are not willing to pay 3x times a sub. But, and this is more on the Revolt marketing guys, I am willing to condider more subs if I can have discounts when purchasing more, as happens in other places with license packs.

The second is, looking at other's needs, as "one identity, multiple people behind" as example the support staff of a given company, that have the "support" account but different people can be on call at a given time, I think a nice implementation can be the ability of "posting a message in behalf of", where an account can enable another account to post messages in its behalf. Ability can be enabled and revoked, and things seems will be quite easy and clean to maintain.
It is the same thing we see in some support sites and emails, we receive mails from support@acme.com but each one has the operator who sent it, or the chat messages with sender "Greg (ACME Support)" in the "ACME Support" chat.
I don't pretend to say how to do it but I see it is possible.

**Reply by gdude2002** [reply]
This doesn't solve the problem as presented:

- Multiple accounts do not scale - One of the reasons I made this suggestion to begin with was to allow plural systems to properly integrate into the platform. These systems can sometimes have hundreds of members, and they can also sometimes be unstable, frequently requiring removing, updating and adding identities.
- Separate contexts can be detrimental - Many plural systems share their memories and social groups among all of their members, and setting up and maintaining a set of accounts for each member makes a lot of things difficult. For example, instead of joining a server on one account, now we need to join it with six!
- Complex use-cases aren't accounted for - Plural systems don't always have a single member "in front", and it's often useful to be able to specify multiple system members together when attributing the authors of a message.

While your approach solves some problems for non-plural folks, the feature as specified was designed with plural systems in mind, and the fact that it'd be useful for other people is a secondary upside.

If you're not familiar with plural systems, feel free to take a look at these sites: https://moderation.wiki/a/plurality and https://morethanone.info/

**Reply by gdude2002** [reply]
Oh, I forgot to mention, but I don't see why this couldn't still factor into a premium model - Revolt could certainly limit this to a certain number of identities, with the option to pay to raise the limit!

**Reply by and-rea** [reply]
> This doesn't solve the problem as presented

Of course it doesn't, I only added some thoughts worth considering.

> If you're not familiar with plural systems, feel free to take a look at these sites: https://moderation.wiki/a/plurality and https://morethanone.info/

Thank you for the links, they helped me focus more your issue. And yes, plural system may also help with my secondary upside.

**Reply by onpon4** [reply]
Having multiple accounts as a plural system would be completely untenable for someone like me as a plural system on a real-time chat app, or anywhere really

Consider DMs with friends, rather than just being able to message my account to contact me, they'd have the extra headache of having to predict which of us is fronting, and I'd have the hassle of looking at 2 separate DM lists, and then when my headmate fronts (which can very easily happen in the middle of a conversation) we'd have to start splitting up a single conversation into multiple different locations, possibly several times

🕷️

**Comment by ShiningVenus** [comment]
No I am all for this for roleplayers. I use Discord for roleplaying and I'd love a system or bot which is like Tupper or even does it better.

**Reply by DeclanChidlow** [reply]
Revolt has a masquerade feature which handles this quite nicely. There is a Revite plugin and plenty of bots that make it easy to use this functionality.

**Reply by gdude2002** [reply]
This is, in fact, the thread that resulted in Masquerade haha 

It'd be nice to see it supported directly in the client though

**Reply by ShiningVenus** [reply]
> Revolt has a masquerade feature which handles this quite nicely. There is a Revite plugin and plenty of bots that make it easy to use this functionality.

Oooh okay then. I figured there'd be a bot or some such for this.

**Comment by scryer-of-static** [comment]
Prior to the Stoat rebranding, I discovered a user called [LazyCat2](https://github.com/LazyCat2) who made a couple plugins, including one for a feature called [Masquerade](https://github.com/LazyCat2/revolt-plugins/tree/main/masquerade). It was incredibly useful, and relatively easy to set up (very clear and simple directions), and it followed your account rather than a bot for a server like in Discord. The caveat is that it only worked for desktop, since the desktop client was the only one that had plugin capability. Also, the new Stoat client doesn't support plugins yet, though when it does I would be interested to find out if this works there too.

**Reply by RianovRV1** [reply]
I cant seem to get masquerade to work on the current Desktop app as of Feb 9th, Ill be joining discussions in stoat to get some help soon but for now was wondering if you or anyone else had some info from this thread

**Reply by gdude2002** [reply]
As far as I can tell, the current Stoat client hasn't progressed a ton over the last few months. It would be excellent to see Masquerade finally exposed in the UI - Perhaps someone out there could PR it?

**Reply by scryer-of-static** [reply]
@RianovRV1 -- The current desktop app (Stoat) does not have a working plugin feature (as far as I'm aware). The plugin I mentioned only worked for the old app (Revolt). I don't believe that app is available for download on the website anymore (though I still have an instance of it). 

@gdude2002 -- Stoat recently got news coverage when Discord announced its new age verification requirements, and Stoat gained a huge influx of users that they weren't prepared for scale-wise. They've since pulled in new devs and made some server changes to accommodate the influx, and they're currently focusing on back-end foundational/stability updates to prep everything for scalability. Currently their focus seems to be the Webapp and (secondarily) the desktop app. It's gotten a few updates since your comment. I get these updates from their dev server on Stoat, as that's the most reliable and most updated place (they don't seem to update any changelogs or blogs right now). 
 
For the new Stoat desktop client (and webapp), there's bots. So far users have found "Masquarade" and "Titan". You can find them [here](https://stt.gg/discover/search?query=masquerade&type=bots). They're very similar to PluralKit and Tupperbox offer a proxy-style setup. Edit: Titan has a way to import PK/TB, though I haven't tested it myself yet.

**Comment by pigsflew** [comment]
User-based sub-profiles and role-based server messaging profiles that one accesses from a drop-down on the message box would be incredible.

It's be really cool if server role profile messages would show "`user` on behalf of `server profile`", either inline or by clicking to view the profile, so that if a particular user's decisions as a mod or whatever are later challenged, there's an indicator of who said what.

For plural folks, I'd also lean towards language like "Sub-profile of `main account profile`" in the more info? But having the messages linked to the system seems a good idea regardless, because we all must be responsible for any actions of our system where they interact with others.

looking forward to seeing this added to the client (web and mobile!)

**Comment by viiuan** [comment]
> a drop-down on the message box

this is exactly what the existing [plugin](https://github.com/LazyCat2/revolt-plugins/tree/main/masquerade) for old.stoat.chat does right now, and something like this being implemented as a native feature for regular users (rather than just bots) would be great, especially since right now using bots for Masquerading as a plural system means the Masquerade is linked to the bot rather than the user who sent it (if you use that plugin, it will open YOUR profile instead).
it would also be nice, as development continues, to actually have info attributable to Masquerades separate from the main account but still have the main account linked. kind of like a per-message server profile... message profile? lmao but basically what PluralChum for BetterDiscord currently does, just more streamlined.

edit: i see [scryer-of-static](https://github.com/orgs/stoatchat/discussions/76#discussioncomment-14872816) already posted about this.

**Reply by Taureon** [reply]
i wonder what the difference in user experience would be as a drop down in this thing

<img width="246" height="368" alt="image" src="https://github.com/user-attachments/assets/922601cd-10a2-4938-8e12-ec7fd9961486" />

considering co-fronting exists, it would probably make more sense for it to be "around here" instead, for ease of access on not just on desktop but also on mobile so mobile users who switch subprofiles often wont be very annoyed, especially during active conversations

<img width="318" height="75" alt="image" src="https://github.com/user-attachments/assets/c2530036-a009-48bb-b853-a247892023f0" />

(singlet here btw in case its relevant, also i dont rp often)

**Comment by MyriaBailey** [comment]
With the mention of "drop-down" menus, I want to make it known that I would find that approach _incredibly_ clunky and annoying. There are situations where I need to utilize multiple different proxies in a row, back to back, in the same conversation, and I do _not_ want to have to leave the text box to go sift through a drop-down menu (or even text search) for rapid back-and-forth proxying.

It could be cool and neat for other use cases, and I'm not opposed to the idea _in general_, but I would find it to be a massive downgrade if I am unable to simply proxy messages directly in-chat via text brackets like what PK/Tupper currently does tbh.

(custom slash commands like what /plu/ral uses could also be permissible, if also kinda clunky still)

**Comment by RianovRV1** [comment]
slash / commands should of course be accepted for ease of access. the dropdown doubles as a handy reference list but shouldnt be the ONLY way to do it I agree

**Reply by MyriaBailey** [reply]
Yeah. I mention slash commands as a middle ground, but they do feel somewhat clunky and awkward when I have a lot of similar, but distinct profiles like ArtoriaSaber, ArtoriaSaberAlter, ArtoriaLancer, ArtoriaLancerAlter, ArtoriaLancerAlterNew, etc. and I don't want to type out the full name, nor use abbreviations (like AS or ALA) when I have so many profiles that would easily start to overlap (approx 170?).

For reference, my typical use case involves sending messages like `. this .` at simplest, or `-0) this (0-` at most complex, and I know many others rely on emojis or other oddities for their proxies as well. That said, I also come from a time before Discord _or_ PluralKit/Tupper, so I'm a bit more old-fashioned here perhaps.

That said, would still be pretty cool to have drop-downs for the purpose of more persistent account profiles, though, even if too slow and bulky for on-the-fly switching like what I would be wanting out of a native Stoat feature here.

**Reply by Taureon** [reply]
> That said, I also come from a time before Discord _or_ PluralKit/Tupper, so I'm a bit more old-fashioned here perhaps.

I wonder what was before that?

**Reply by selenated** [reply]
> > That said, I also come from a time before Discord _or_ PluralKit/Tupper, so I'm a bit more old-fashioned here perhaps.
> 
> I wonder what was before that?
> 

Before that people would just use "brackets" (what PK and Tbox use for proxying already) or sign-offs to indicate who was speaking. `:: Something like this::` for an example of brackets

Source: was also around before proxy bots, plus I used to frequent places where proxies aren't really possible (Reddit, forums, other chatrooms, etc)

**Reply by MyriaBailey** [reply]
Nothing! Not to my knowledge, at least. I was on Kik and Kakao before Discord, which didn't have bots like that as far as I knew. Instead of bot-proxying, we just used text brackets and learned to recognize who was who like a sort of text-based facial recognition, and chose brackets accordingly :P

Other spaces I was in would also just change username/pfps altogether though, but that only really worked when you were only RP'ing for one character at a time.

**Comment by rarespawns** [comment]
Genuinely this would be an incredible addition!
However, the main issue I see is that the dropdown menu could get clunky fast with larger amounts of proxies, there should be slash commands, and possibly also a proxy tag system similar to pluralkit and tupperbox from Discord where you have something added to your message ("something[text], [text]something, and some[text]thing) to send a message as someone, and maaaybe also something like latch autoproxy as a setting, where it just automatically proxies as the last proxied member

**Reply by Taureon** [reply]
i think for the large dropdown lists it could be somewhat mitigated by using 2d lists (so grids), maybe even clone the emoji picker and add categories and a small search bar

of course still add other means to do the same thing, just wanna make sure those who prefer a specific way arent screwed
